internal/controller/gateway.go

```diff
@@ -189,11 +196,150 @@ func (c *GatewayController) ensureExtensionPolicy(ctx context.Context, gw *gwapi
+	// Collect all unique InferencePool references from this route.
+	inferencePoolEPPs := make(map[string]*gwaiev1a2.InferencePool)
+
+	for _, rule := range route.Spec.Rules {
+		if rule.HasInferencePoolBackends() {
+			for _, backendRef := range rule.BackendRefs {
+				if backendRef.IsInferencePool() {
+					// Get the InferencePool resource to extract EPP configuration.
+					var inferencePool gwaiev1a2.InferencePool
+					err := c.client.Get(ctx, client.ObjectKey{
+						Namespace: route.Namespace,
+						Name:      backendRef.Name,
+					}, &inferencePool)
+					if err != nil {
+						return fmt.Errorf("failed to get InferencePool %s/%s: %w", route.Namespace, backendRef.Name, err)
+					}
+
+					// Use a unique key for each InferencePool.
+					key := fmt.Sprintf("%s/%s", route.Namespace, backendRef.Name)
+					inferencePoolEPPs[key] = &inferencePool
+				}
+			}
+		}
+	}
+
+	// Create EnvoyExtensionPolicy for each unique EPP service, targeting the HTTPRoute.
+	for key, inferencePool := range inferencePoolEPPs {
+		if err := c.ensureEPPExtensionPolicyForHTTPRoute(ctx, gw, route, key, inferencePool); err != nil {
+			return fmt.Errorf("failed to ensure EPP extension policy for %s: %w", key, err)
+		}
+	}
+
+	return nil
+}
+
+// ensureEPPExtensionPolicyForHTTPRoute creates an EnvoyExtensionPolicy for a specific InferencePool's EPP service, targeting an HTTPRoute.
+func (c *GatewayController) ensureEPPExtensionPolicyForHTTPRoute(ctx context.Context, gw *gwapiv1.Gateway, route *aigv1a1.AIGatewayRoute, inferencePoolKey string, inferencePool *gwaiev1a2.InferencePool) error {
+	if inferencePool.Spec.ExtensionRef == nil || inferencePool.Spec.ExtensionRef.Name == "" {
+		// No EPP service configured, skip.
+		return nil
+	}
+
+	eppServiceName := string(inferencePool.Spec.ExtensionRef.Name)
+	var eppServicePort int32 // Default port.
+	if inferencePool.Spec.ExtensionRef.PortNumber != nil {
+		eppServicePort = int32(*inferencePool.Spec.ExtensionRef.PortNumber)
+	} else {
+		// The port number on the service running the extension. When unspecified,
+		// implementations SHOULD infer a default value of 9002 when the Kind is
+		// Service.
+		eppServicePort = 9002
+	}
+
+	// Create Backend resource for the EPP service.
+	backendName := fmt.Sprintf("epp-%s-%s", route.Name, strings.ReplaceAll(inferencePoolKey, "/", "-"))
+	backend := egv1a1.Backend{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backendName,
+			Namespace: gw.Namespace,
+		},
+		Spec: egv1a1.BackendSpec{
+			Endpoints: []egv1a1.BackendEndpoint{{
+				FQDN: &egv1a1.FQDNEndpoint{
+					Hostname: fmt.Sprintf("%s.%s.svc.cluster.local", eppServiceName, route.Namespace),
+					Port:     eppServicePort,
+				},
+			}},
+		},
+	}
+
+	if err := c.client.Create(ctx, &backend); err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create EPP backend: %w", err)
+	}
```

Comment:
We need to handle the update case as well as InferencePool resource can be updated like FQDN or port.

Reply:
instead of creating eep, how about adding extproc into the specific route which routes to the inference pool lb policy cluster? that why other normal routes won't need to talk to eep unnecessarily. This could be done in extension server i guess?